pwlwp/matcher.py

```diff
@@ -67,36 +89,106 @@ def __str__(self) -> str:
 	
 class Matcher():
 
-	def __init__(self, sentence_transformer, data_embeddings, data_frame: DataFrame, threshold: float) -> None:
+	def __init__(self, sentence_transformer, data_embeddings, data_frame: DataFrame, threshold: float, threshold2: float) -> None:
 		super().__init__()
 		self.sentence_transformer = sentence_transformer
 		self.data_frame = data_frame
 		self.threshold = threshold
+		self.threshold2 = threshold2
 		self.data_embeddings = data_embeddings
 		self.causal_column = self.data_frame["causal"]
 		self.belief_column = self.data_frame["belief"]
 	
-	def match_choice(self, choice: str) -> ChoiceMatch:
-		choice_embedding = self.sentence_transformer.encode(choice)
-		sentence_matches = [
-			SentenceMatch(
-				choice_embedding,
-				self.data_embeddings[index],
-				self.threshold,
-				is_causal=bool(self.causal_column[index]),
-				is_belief=bool(self.belief_column[index])
-			)
-			for index in range(len(self.data_frame))
-		]
+	def match_choice(self, introduction: str, choice: str, to_print: bool, filter_first: bool) -> ChoiceMatch:
+
+		sentence_matches = []
+
+		if not filter_first:
+
+			initial_embedding = self.sentence_transformer.encode(choice)
+			introduction_embedding = self.sentence_transformer.encode(introduction)
+			choice_embedding = numpy.add(initial_embedding, introduction_embedding) / 2
+
+			sentence_matches = [
+				SentenceMatch(
+					choice,
+					choice_embedding,
+					self.data_embeddings[index],
+					self.threshold,
+					is_causal=bool(self.causal_column[index]),
+					is_belief=bool(self.belief_column[index])
+				)
+				for index in range(len(self.data_frame))
+			]
+
+		else:
+
+			initial_embedding = self.sentence_transformer.encode(choice)
+			introduction_embedding = self.sentence_transformer.encode(introduction)
+
+			sentence_matches = [
+				SentenceMatchFilter(
+					choice,
+					introduction_embedding,
+					initial_embedding,
+					self.data_embeddings[index],
+					self.threshold,
+					self.threshold2,
+					is_causal=bool(self.causal_column[index]),
+					is_belief=bool(self.belief_column[index])
+				)
+				for index in range(len(self.data_frame))
+			]
+
+		if to_print:
```

Comment:
Would this print all the same stuff with a single `if to_print:` at the beginning?

Reply:
Yes that is true. Initially I was doing some processing alongside the printing, that's why it's structured that way. I'll fix it in a bit.

Reply:
No need yet.  There will probably be other things to change.  Thanks for checking.